[Frontend] Support simpler image input format

yue-anyscale opened the pull request:
**Why we make this change:**
Some users use a simpler image input format (example below), which we cannot handle properly now:
```
[
  {
    "role": "user",
    "content": [
      "what's the animals in the image?",
      {
        "image_url": "https://animals.com/dog"
      },
      {
        "image_url": "https://animals.com/cat"
      },
      
    ]
  }
]
```

OpenAI can handle this format well even though they didn't document in the API reference. 

**Changes:**
Modified the input parsing logic to support more content types within the "content" field - string, simpler image_url, simpler audio_url.

yue-anyscale committed:
support more simple image input format -- which is supported by open ai but not show in public doc

yue-anyscale committed:
nit

yue-anyscale committed:
handle simple audio input format as well

yue-anyscale committed:
typing

comaniac commented on vllm/entrypoints/chat_utils.py:
```diff
@@ -57,11 +57,32 @@ class CustomChatCompletionContentPartParam(TypedDict, total=False):
     """The type of the content part."""
 
+class CustomChatCompletionContentSimpleImageParam(TypedDict, total=False):
+    """A simpler version of the param that only accepts a plain image_url.
```
```suggestion
    """A simpler version of the param that only accepts a plain image_url.
    This is supported by OpenAI API, although it is not documented.
```

yue-anyscale committed:
format

comaniac commented on vllm/entrypoints/chat_utils.py:
```diff
@@ -401,29 +437,42 @@ def _parse_chat_message_content_parts(
+            mm_parser.parse_image(part["image_url"])  # type: ignore
             has_image = True
-        elif part_type == "audio_url":
-            audio_url = _AudioParser(part)["audio_url"]
-
-            mm_parser.parse_audio(audio_url["url"])
-        elif part_type == "refusal":
-            text = _RefusalParser(part)["refusal"]
-            texts.append(text)
+        elif _is_simple_audio_part(part):
+            mm_parser.parse_audio(part["audio_url"])  # type: ignore
```
I'm thinking about a more concise way like the following:

```python
if isinstance(part, str):
    text = _TextParser(part)
    texts.append(text)
else:
    assert isinstance(part, dict)
    part_type, part_url = parse_mm_part()
    
    if part_type == "text":
        ...
    else:
        if part_type == "image_url":
            mm_parser.parse_image(part_url)
            has_image = True
        elif part_type == "audio_url":
            ...
```

```python
def parse_mm_part(part: ChatCompletionContentPartParam):
    part_type = part.get("type", None)
    if part_type == "image_url":
        image_url = _ImageParser(part)["image_url"]
        if image_url.get("detail", "auto") != "auto":
            logger.warning(
                "'image_url.detail' is currently not supported and "
                "will be ignored.")
        url = image_url["url"]
    elif part_type == "audio_url":
        ...
    elif part_type is None:
        if "image_url" in part and isinstance(part["image_url"], str):
            part_type = "image_url"
            url = part["image_url"]
        elif ...
    else:
        ValueError(...)
    return part_type, url
```

WDYT?

yue-anyscale replied:
this makes more sense! let me do this

comaniac commented:
Overall LGTM. Also please add a unit test in `tests/entrypoints/test_chat_utils.py`

Also cc @ywang96 @DarkLight1337

yue-anyscale committed:
address comments - will add unittest

yue-anyscale committed:
format

yue-anyscale committed:
add unittest for simple image input format

yue-anyscale commented:
> Overall LGTM. Also please add a unit test in `tests/entrypoints/test_chat_utils.py`
> 
> Also cc @ywang96 @DarkLight1337

Thanks for review! I just addressed comments and added unittest.

comaniac commented on vllm/entrypoints/chat_utils.py:
```diff
@@ -386,6 +411,57 @@ def _get_full_multimodal_text_prompt(placeholder_counts: Dict[str, int],
+    Raises:
+        ValueError: If the 'type' field is missing and no direct URL is found.
+    """
+    part_type = part.get("type", None)  # type: ignore
```
Do not ignore type if it can be easily fixed. In this case, I guess the following change should fix the type error
```python
part_type: Optional[str] = part.get("type", None)
```
Plz also try to fix other types.

yue-anyscale replied:
sounds good, used proper way addressed this, no type ignore now.

comaniac commented on vllm/entrypoints/chat_utils.py:
```diff
@@ -401,29 +477,22 @@ def _parse_chat_message_content_parts(
+            assert isinstance(part, dict)
+            part_type, content = _parse_chat_message_content_mm_part(part)
+
+            if part_type in ["text", "refusal"]:
```
Use tuple instead of list is a better practice in terms of performance for a static list.

```suggestion
            if part_type in ("text", "refusal"):
```

comaniac commented on vllm/entrypoints/chat_utils.py:
```diff
@@ -386,6 +411,57 @@ def _get_full_multimodal_text_prompt(placeholder_counts: Dict[str, int],
+
+    # Handle missing 'type' but provided direct URL fields.
+    if part_type is None:
+        for url_type in ["image_url", "audio_url"]:
```
ditto

comaniac reviewed (approved):
Overall LGTM. Please fix the type errors instead of ignoring them.
After that I'll put ready to unblock the rest CI tests.

yue-anyscale committed:
address comments - better handle types, remove the ignore types, and use tuple instead of list

yue-anyscale committed:
nits

yue-anyscale commented:
> Overall LGTM. Please fix the type errors instead of ignoring them. After that I'll put ready to unblock the rest CI tests.

Thanks for the reviews! Fixed.

DarkLight1337 commented on vllm/entrypoints/chat_utils.py:
```diff
@@ -386,6 +411,71 @@ def _get_full_multimodal_text_prompt(placeholder_counts: Dict[str, int],
+
+    if not isinstance(part_type, str):
+        raise ValueError("Invalid 'type' field in multimodal part.")
+    return part_type, "unknown part_type content"
```
I think we should throw an error in this case instead of silently overwriting the content.

yue-anyscale replied:
thanks for review! The error will be throw in the caller if part_type is unknown https://github.com/vllm-project/vllm/pull/9478/files#diff-31e6bd0df09a47b5587701203d558701ac46e4f85bf7db83632da9990eaef198R515
